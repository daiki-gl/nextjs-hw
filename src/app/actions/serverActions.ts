'use server'

export async function getUsers() {
  const res = await fetch(
    'https://gist.githubusercontent.com/daiki-gl/4fcf6cb5f0080ed487b254f4fd6eec35/raw/48956c1c83b6d1618e3ed65feddad5ad97b4008d/user.json',
    {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    }
  )
  const data = await res.json()

  return data
}
