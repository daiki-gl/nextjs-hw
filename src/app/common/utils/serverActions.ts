'use server'

/* ユーザーデータを取得するAPI */
export async function getUsers() {
  const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL

  // IF0001.jsonから値取得
  const res = await fetch(`${BASE_URL}/api/stub/IF0001.json`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
  })
  const data = (await res).json()
  return data
}
