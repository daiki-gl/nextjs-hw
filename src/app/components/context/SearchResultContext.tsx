'use client';
import { SearchResultContextType, UserData } from '@/app/common/types';
import { createContext, useContext, useState } from 'react';



const SearchResultContext = createContext<SearchResultContextType | null>(null);

export const SearchResultProvider = ({ children }: { children: React.ReactNode }) => {
  const [searchResult, setSearchResult] = useState<UserData[]>([]);

  return (
    <SearchResultContext.Provider value={{ searchResult, setSearchResult }}>
      {children}
    </SearchResultContext.Provider>
  );
};

export const useSearchResultContext = () => {
  const context = useContext(SearchResultContext);
  if (!context) throw new Error('useShowDataContext must be used within ShowDataProvider');
  return context;
};