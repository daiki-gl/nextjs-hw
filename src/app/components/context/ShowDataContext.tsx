'use client';
import { createContext, useContext, useState } from 'react';
import { ShowDataContextType, UserData } from '../../common/types';


const ShowDataContext = createContext<ShowDataContextType | null>(null);

export const ShowDataProvider = ({ children }: { children: React.ReactNode }) => {
  const [showAddUserData, setShowAddUserData] = useState<UserData[]>([]);

  return (
    <ShowDataContext.Provider value={{ showAddUserData, setShowAddUserData }}>
      {children}
    </ShowDataContext.Provider>
  );
};

export const useShowDataContext = () => {
  const context = useContext(ShowDataContext);
  if (!context) throw new Error(' must be used within ShowDataProvider');
  return context;
};