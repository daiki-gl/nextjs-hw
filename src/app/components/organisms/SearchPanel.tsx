'use client'
import { SetStateAction, useEffect } from "react";
import Form from "../molecules/Form";
import useUserList from "@/app/common/hooks/useUserList";
import { UserData } from "@/app/common/types";
import Modal from "../molecules/Modal";
import useModal from "@/app/common/hooks/useModal";

export default function SearchPanel({showUserData, setShowUserData}
    : {
        showUserData: UserData[] | null; 
        setShowUserData: React.Dispatch<SetStateAction<UserData[]>>;
    }) {

    const {isOpen, openModal, closeModal} = useModal();
    const {getUsersFunction,users} = useUserList();
    useEffect(() => {
        (async () => {
        await getUsersFunction();
        })()
    },[]);

    return (
        <div>
            <Form showUserData={showUserData} setShowUserData={setShowUserData} userData={users} onOpen={openModal} />
            <Modal isOpen={isOpen} onClose={closeModal} />
        </div>
    )
}