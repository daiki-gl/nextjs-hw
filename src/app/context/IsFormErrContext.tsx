'use client';
import { createContext, useContext, useState } from 'react';
import { IsFormErrContextType } from '../common/types';


const IsFormErrContext = createContext<IsFormErrContextType | null>(null);

export const IsFormErrContextProvider = ({ children }: { children: React.ReactNode }) => {
  const [isFormErr, setIsFormErr] = useState<boolean>(true);

  return (
    <IsFormErrContext.Provider value={{ isFormErr, setIsFormErr }}>
      {children}
    </IsFormErrContext.Provider>
  );
};

export const useIsFormErrContext = () => {
  const context = useContext(IsFormErrContext);
  if (!context) throw new Error('useShowDataContext must be used within ShowDataProvider');
  return context;
};