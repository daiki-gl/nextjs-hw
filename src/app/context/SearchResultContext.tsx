'use client';
import { createContext, useContext, useState } from 'react';
import { SearchResultContextType, UserData } from '../common/types';


const SearchResultContext = createContext<SearchResultContextType | null>(null);

export const SearchResultProvider = ({ children }: { children: React.ReactNode }) => {
  const [searchResult, setSearchResult] = useState<UserData[]>([]);

  return (
    <SearchResultContext.Provider value={{ searchResult, setSearchResult }}>
      {children}
    </SearchResultContext.Provider>
  );
};

export const useSearchResultContext = () => {
  const context = useContext(SearchResultContext);
  if (!context) throw new Error('useShowDataContext must be used within ShowDataProvider');
  return context;
};