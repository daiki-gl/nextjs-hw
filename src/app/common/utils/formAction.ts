import React, { SetStateAction } from 'react'
import { UserData } from '../types'
import { getUsers } from './serverActions'

/*
 Formのsubmitに関わる関数が格納されている
 @param event formのevent
 @param userData APIで取得したユーザー情報一覧
 @param setShowUserData 表示用のユーザーをsetする関数
*/
const startIndex = 0 // 1ページ目の開始インデックス
const endIndex = startIndex + 10 // 1ページ目の終了インデックスs

export const handleSubmit = async (
  event: React.FormEvent<HTMLFormElement>,
  userData: UserData[] | null,
  setShowUserData: (value: React.SetStateAction<UserData[]>) => void,
  setSearchResult: React.Dispatch<React.SetStateAction<UserData[]>>
) => {
  /*
  フォーム送信時のユーザー検索処理
  入力された名前・電話番号・住所・支払い情報で userData をフィルタリングして setShowUserData に反映。
  */
  event.preventDefault()
  const form = event.currentTarget

  const users = await getUsers().then((data) => {
    return data
  })

  /* 各インプット要素を取得 */
  // const form = event.currentTarget
  const nameInput = form.elements.namedItem('name') as HTMLInputElement
  const phoneInput = form.elements.namedItem('phone') as HTMLInputElement
  const addressInput = form.elements.namedItem('address') as HTMLInputElement
  const paymentFromInput = form.elements.namedItem(
    'payment-from'
  ) as HTMLInputElement
  const paymentToInput = form.elements.namedItem(
    'payment-to'
  ) as HTMLInputElement

  /* 各入力値取得 */
  const nameValue = nameInput.value
  const phoneValue = phoneInput.value
  const addressValue = addressInput?.value
  const paymentFromValue = paymentFromInput?.value
  const paymentToValue = paymentToInput?.value

  // ユーザーデータが存在する場合フィルタリング
  if (userData) {
    // const filteredData = userData.filter((user: UserData) => {
    //   const filters = [
    //     { value: nameValue, match: user.name.includes(nameValue) },
    //     { value: phoneValue, match: user.phone.includes(phoneValue) },
    //     { value: addressValue, match: user.address.includes(addressValue) },
    //     {
    //       value: paymentFromValue,
    //       match: user.payment.includes(paymentFromValue),
    //     },
    //     {
    //       value: paymentToInput,
    //       match: user.payment.includes(paymentToValue),
    //     },
    //   ]

    //   // 入力されたフィールドがすべて一致するかどうかをチェック
    //   return filters
    //     .filter((f) => f.value) // 入力された値だけを対象にする
    //     .every((f) => f.match) // すべての条件が一致するか確認
    // })

    localStorage.setItem(
      'searchValue',
      JSON.stringify({
        name: nameValue,
        phone: phoneValue,
        address: addressValue,
        paymentFrom: paymentFromValue,
        paymentTo: paymentToValue,
      })
    )

    // 1ページ目に表示するデータをセット
    setShowUserData(users.slice(startIndex, endIndex))
    // setShowUserData(filteredData.slice(startIndex, endIndex))
    setSearchResult(users)
    return
  }
}

/*
 フォームの入力変更時に、状態を更新する。
 数値と文字列の型に応じて適切な setState を使い分ける。
 */
export const handleInputChange = (
  e: React.ChangeEvent<HTMLInputElement>,
  setInput:
    | React.Dispatch<React.SetStateAction<string>>
    | React.Dispatch<React.SetStateAction<number>>
) => {
  if (typeof e.target.value === 'string') {
    ;(setInput as React.Dispatch<React.SetStateAction<string>>)(e.target.value)
  }
}

/*
 入力値の正規表現チェック。
 - 正しい形式 → エラー解除
 - 空文字 → 入力エラーなし・フォームエラーあり
 - 不正な形式 → 入力エラーあり・フォームエラーあり
 */
export const handleErrCheck = (
  regex: RegExp,
  inputData: string,
  setInputErr: React.Dispatch<SetStateAction<boolean>>,
  setIsFormErr: React.Dispatch<SetStateAction<boolean>>
) => {
  const isValid = regex.test(inputData)
  setInputErr(!isValid) // 項目個別のエラー更新

  // フォーム全体のエラーを更新（※ここでは簡易的に）
  setIsFormErr(!isValid)
}

export const handleSubmitted = (
  setShowUserData: (value: React.SetStateAction<UserData[]>) => void,
  setSearchResult: React.Dispatch<React.SetStateAction<UserData[]>>,
  searchResult: UserData[]
) => {
  let usersToAdd: UserData[] = []
  if (typeof window !== 'undefined') {
    const storedUsersToAdd = localStorage.getItem('usersToAdd')
    if (storedUsersToAdd) {
      try {
        usersToAdd = JSON.parse(storedUsersToAdd)
      } catch (e) {
        console.error("Failed to parse 'usersToAdd' from localStorage:", e)
        localStorage.removeItem('usersToAdd') // パースエラーの場合は削除
      }
    }
  }

  // 除外するユーザーのIDをSetに格納
  const usersToAddIds = new Set(usersToAdd.map((user) => user.id))
  const filteredUsers = searchResult.filter(
    (user) => !usersToAddIds.has(user.id)
  )
  // フォーム送信後の処理
  setShowUserData(filteredUsers.slice(startIndex, endIndex))
  setSearchResult(filteredUsers)
}
