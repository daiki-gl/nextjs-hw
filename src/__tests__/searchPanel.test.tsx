import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import SearchPanel from '../app/components/organisms/SearchPanel';
import { getUsers } from '../app/common/utils/serverActions';
import { IsFormErrContextProvider, useIsFormErrContext } from '../app/components/context/IsFormErrContext';
import { TypeContextProvider } from '../app/components/context/TypeContext';
import { SearchResultProvider } from '../app/components/context/SearchResultContext';
import { ShowDataProvider } from '../app/components/context/ShowDataContext';
import { handleSubmit } from '../app/common/utils/formAction';

const mockUsers = [
  { id: 1, name: '山田太郎', phone: '09012345678', address: '東京都', status: false, payment: '1000' },
  { id: 2, name: '佐藤花子', phone: '08098765432', address: '大阪府', status: false, payment: '2000' },
  { id: 3, name: '田中一郎', phone: '07011112222', address: '神奈川県', status: true, payment: '3000' },
];

// getUsersをモック
jest.mock('../app/common/utils/serverActions', () => ({
  getUsers: jest.fn(() => Promise.resolve(mockUsers)),
}));

// formActionのhandleSubmitをモック
jest.mock('../app/common/utils/formAction', () => ({
  handleSubmit: jest.fn(e => e.preventDefault()),
  handleInputChange: jest.fn(),
}));

jest.mock('../app/components/context/IsFormErrContext', () => ({
  ...jest.requireActual('../app/components/context/IsFormErrContext'),
  useIsFormErrContext: jest.fn(),
}));

const mockSetIsFormErr = jest.fn();
(useIsFormErrContext as jest.Mock).mockReturnValue({
  isFormErr: {}, // デフォルトで空のオブジェクトを返す
  setIsFormErr: mockSetIsFormErr,
});

describe('SearchPanel', () => {
const renderComponent = (type = "list") => {
Object.defineProperty(window, 'sessionStorage', {
  value: {
    getItem: jest.fn().mockReturnValue(type),
    setItem: jest.fn(),
    removeItem: jest.fn(),
    clear: jest.fn(),
  },
  writable: true,
});
   render(
      <IsFormErrContextProvider>
        <TypeContextProvider>
          <SearchResultProvider>
            <ShowDataProvider>
              <SearchPanel showUserData={null} setShowUserData={function (): void {
                            throw new Error('Function not implemented.');
                        } } />
            </ShowDataProvider>
          </SearchResultProvider>
        </TypeContextProvider>
      </IsFormErrContextProvider>
    );
  };

  it('初期ロード時にLoading...と表示され、その後フォームが表示される', async () => {
    (getUsers as jest.Mock).mockImplementationOnce(() => new Promise(resolve => setTimeout(() => resolve(mockUsers), 100)));
    renderComponent();
    await waitFor(() => {
      expect(screen.getByLabelText('名前')).toBeInTheDocument();
      expect(screen.getByLabelText('電話番号')).toBeInTheDocument();
      expect(screen.getByLabelText('郵便番号')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'help' })).toBeInTheDocument();
      const searchButton = screen.getByRole('button', { name: '検索' });
      expect(searchButton).toBeInTheDocument();
      expect(searchButton).toBeDisabled();
    });
  });

  it('フォームが送信されたときにhandleSubmitが呼び出されること', async () => {
    renderComponent();
    await waitFor(() => {
      const nameInput = screen.getByLabelText('名前');
      const form = screen.getByRole('form');
      expect(form).toBeInTheDocument();
      fireEvent.change(nameInput, { target: { value: 'テスト太郎' } });
      fireEvent.submit(form);
      // handleSubmitが呼び出されたことを確認
      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });
  });

  it('スナップショットテスト', async () => {
    const { asFragment } = render(
      <IsFormErrContextProvider>
        <TypeContextProvider>
          <SearchResultProvider>
            <ShowDataProvider>
              <SearchPanel showUserData={null} setShowUserData={function (): void {
                            throw new Error('Function not implemented.');
                        } } />
            </ShowDataProvider>
          </SearchResultProvider>
        </TypeContextProvider>
      </IsFormErrContextProvider>
    );
    await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
  });

  it('sessionStorageのsearchTypeがlistの時、郵便番号の入力フィールドが表示される', async () => {
    renderComponent("list");
    await waitFor(() => {
      expect(screen.getByLabelText('郵便番号')).toBeInTheDocument();
    });
  });

  it('sessionStorageのsearchTypeがdetailの時、FromとToの金額の入力フィールドが表示される', async () => {
    renderComponent("detail");
    await waitFor(() => {
      expect(screen.getByLabelText('From')).toBeInTheDocument();
      expect(screen.getByLabelText('To')).toBeInTheDocument();  
    });
  });
  
  it('searchTypeがdetailの時のスナップショットテスト', async () => { 
      Object.defineProperty(window, 'sessionStorage', {
          value: {
              getItem: jest.fn().mockReturnValue("detail"),
        setItem: jest.fn(),
        removeItem: jest.fn(),
        clear: jest.fn(),
      },
      writable: true,
    });
    const { asFragment } = render(
      <IsFormErrContextProvider>
        <TypeContextProvider>
          <SearchResultProvider>
            <ShowDataProvider>
              <SearchPanel showUserData={null} setShowUserData={function (): void {
                            throw new Error('Function not implemented.');
                        } } />
            </ShowDataProvider>
          </SearchResultProvider>
        </TypeContextProvider>
      </IsFormErrContextProvider>
    );
    await waitFor(() => {
    expect(asFragment()).toMatchSnapshot();
  });
  });
});

