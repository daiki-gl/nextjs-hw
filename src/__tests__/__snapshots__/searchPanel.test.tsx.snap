// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SearchPanel searchTypeがdetailの時のスナップショットテスト 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mx-auto w-3/5"
    >
      <form
        action="/search"
        aria-label="form"
        class="flex justify-between flex-wrap"
        method="get"
      >
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="name"
          >
            名前
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="name"
            max="9999999999"
            min="0"
            name="name"
            placeholder="名前入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="phone"
          >
            電話番号
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="phone"
            max="9999999999"
            min="0"
            name="phone"
            placeholder="電話番号入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="shrink-0 w-[785px] pt-2 -mb-3"
        >
          支払い金額
        </div>
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="payment-from"
          >
            From
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="payment-from"
            max="9999999999"
            min="0"
            name="payment-from"
            placeholder="金額入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="payment-to"
          >
            To
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="payment-to"
            max="9999999999"
            min="0"
            name="payment-to"
            placeholder="金額入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="mt-10 mx-auto "
        >
          <button
            class="cursor-pointer  
                
                text-blue-500 hover:text-gray-300 underline inline-block mr-22 text-left hover:!bg-transparent
                
                 hover:bg-blue-600  
                "
            type="button"
          >
            help
          </button>
          <button
            class="cursor-pointer  
                outline-1 w-48 text-left flex-1/2
                
                
                opacity-50 !cursor-not-allowed bg-gray-500 
                "
            disabled=""
            type="submit"
          >
            検索
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchPanel スナップショットテスト 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mx-auto w-3/5"
    >
      <form
        action="/search"
        aria-label="form"
        class="flex justify-between flex-wrap"
        method="get"
      >
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="name"
          >
            名前
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="name"
            max="9999999999"
            min="0"
            name="name"
            placeholder="名前入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="phone"
          >
            電話番号
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="phone"
            max="9999999999"
            min="0"
            name="phone"
            placeholder="電話番号入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="mt-10 basis-1/2 flex-1/2"
        >
          <label
            class="inline-block w-30"
            for="address"
          >
            郵便番号
          </label>
          <input
            class="outline-1 mr-14 w-48"
            id="address"
            max="9999999999"
            min="0"
            name="address"
            placeholder="郵便番号入力欄"
            type="text"
            value=""
          />
          <div>
            <label
              class="mr-[104px]"
            />
          </div>
        </div>
        <div
          class="mt-10 flex-1/2 "
        >
          <button
            class="cursor-pointer  
                
                text-blue-500 hover:text-gray-300 underline inline-block mr-22 text-left hover:!bg-transparent
                
                 hover:bg-blue-600  
                "
            type="button"
          >
            help
          </button>
          <button
            class="cursor-pointer  
                outline-1 w-48 text-left flex-1/2
                
                
                opacity-50 !cursor-not-allowed bg-gray-500 
                "
            disabled=""
            type="submit"
          >
            検索
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
